crates/ide-assists/src/handlers/extract_function.rs

```diff
@@ -955,6 +990,26 @@ impl FunctionBody {
     }
 }
 
+fn parent_where_clauses(parent: &SyntaxNode) -> Vec<ast::WhereClause> {
+    let mut where_clause: Vec<ast::WhereClause> = parent
+        .ancestors()
+        .filter_map(ast::AnyHasGenericParams::cast)
+        .filter_map(|it| it.where_clause())
+        .collect();
+    where_clause.reverse();
+    where_clause
+}
+
+fn parent_generic_param_lists(parent: &SyntaxNode) -> Vec<ast::GenericParamList> {
+    let mut generic_param_list: Vec<ast::GenericParamList> = parent
+        .ancestors()
+        .filter_map(ast::AnyHasGenericParams::cast)
+        .filter_map(|it| it.generic_param_list())
+        .collect();
+    generic_param_list.reverse();
+    generic_param_list
+}
```

Comment:
These are a bit too "weakly typed". There are only really up to two things that can offer use generics, the function this is an expression of (or const/static but those don't currently have generics) and then a trait/impl that contains that. So we should check for those two only.

So `parent.ancestors().find_map(ast::Item::cast)`, then check that this is either an `ast::Fn`, `ast::Static` or `ast::Const` since those are the only valid ones (nothing else should contain expressions atm).

Then with that item at hand we can do `item.syntax().parent().and_then(ast::Item::cast)` and check if thats an `ast::Impl` or `ast::Trait`.

With those two things we can then fetch the generics and where clauses instead. Thats a bit nicer and that way we also dont grab unrelated generics and where clauses

Reply:
Okay, I'll try that out! It wasn't clear to me how specific or general I should be when looking up through the ancestors, I'll try to be more specific in the future.

Can you think up an example of an "unrelated" generic that this approach would capture? If so, I'll add it as a unit test.

Reply:
Ah, I see! I had been thinking that was a case we wanted to catch, but I just tried it out locally and learned about a new compiler error: https://doc.rust-lang.org/error-index.html#E0401

I'll make this change to be more precise now, thanks for the help!

Reply:
Okay, I've tried this now, what do you think?